timeline_sync/api.py

```diff
@@ -0,0 +1,172 @@
+from flask import Blueprint, jsonify, url_for, request
+import datetime
+import secrets
+import uuid
+import requests
+from .models import db, SandboxToken, TimelinePin, UserTimeline, UserSubscription
+from .utils import get_uid, api_error
+from .settings import config
+
+api = Blueprint('api', __name__)
+
+
+def parse_user_token(user_token):
+    if user_token is None:
+        raise ValueError
+    sandbox_token = SandboxToken.query.filter_by(token=user_token).one_or_none()
+    if sandbox_token is not None:
+        return sandbox_token.user_id, sandbox_token.app_uuid, f"sandbox-uuid:{sandbox_token.app_uuid}"
+    else:
+        result = requests.get(f"{config['APPSTORE_API_URL']}/api/v1/locker/by_token/{user_token}")
+        if result.status_code != 200:
+            raise ValueError
+        locker_info = result.json()
+        return locker_info['user_id'], locker_info['app_uuid'], f"uuid:{locker_info['app_uuid']}"
+    # TODO: should dataSource be app_uuid or it's something else?
+
+
+@api.route('/tokens/sandbox/<app_uuid>')
+def get_sandbox_token(app_uuid):
+    uid = get_uid()
+    app_uuid = uuid.UUID(app_uuid)
+
+    sandbox_token = SandboxToken.query.filter_by(user_id=uid, app_uuid=app_uuid).one_or_none()
+
+    if sandbox_token is None:
+        # TODO: return 404 if app does not have timeline support or user is not authorised in dev portal
+        sandbox_token = SandboxToken(user_id=uid, app_uuid=app_uuid, token=secrets.token_urlsafe(32))
+        db.session.add(sandbox_token)
+        db.session.commit()
+
+    result = {"uuid": sandbox_token.app_uuid, "token": sandbox_token.token}
+
+    return jsonify(result)
+
+
+@api.route('/sync')
+def sync():
+    user_id = get_uid()
+
+    user_timeline = db.session.query(UserTimeline).filter_by(user_id=user_id)
+
+    updates = [user_timeline_item.to_json() for user_timeline_item in user_timeline]
+
+    user_timeline.delete()
+    db.session.commit()
+
+    result = {
+        "updates": updates,
+        "syncURL": url_for('api.sync', _external=True)
```

Comment:
I think this results in each call to `/sync` returning every pin the user has ever seen. It would be better to let the apps do an incremental sync. We can probably use the `UserTimeline`'s primary key to do so — inject that into the returned `syncURL`, then query for only `UserTimeline` entries with an ID greater than that one.